prow/cmd/generic-autobumper/main_test.go

```diff
@@ -0,0 +1,843 @@
+			}
+		})
+	}
+}
+
+func TestParseUpstreamImageVersion(t *testing.T) {
+	cases := []struct {
+		description            string
+		upstreamURL            string
+		upstreamServerResponse string
+		expectedRes            string
+		expectError            bool
+		prefix                 string
+	}{
+		{
+			description:            "empty upstream URL will return an error",
+			upstreamURL:            "",
+			upstreamServerResponse: "",
+			expectedRes:            "",
+			expectError:            true,
+			prefix:                 "gcr.io/k8s-prow/",
+		},
+		{
+			description:            "an invalid upstream URL will return an error",
+			upstreamURL:            "whatever-url",
+			upstreamServerResponse: "",
+			expectedRes:            "",
+			expectError:            true,
+			prefix:                 "gcr.io/k8s-prow/",
+		},
+		{
+			description:            "an invalid response will return an error",
+			upstreamURL:            "auto",
+			upstreamServerResponse: "whatever-response",
+			expectedRes:            "",
+			expectError:            true,
+			prefix:                 "gcr.io/k8s-prow/",
+		},
+		{
+			description:            "a valid response will return the correct tag",
+			upstreamURL:            "auto",
+			upstreamServerResponse: "     image: gcr.io/k8s-prow/deck:v20200717-cf288082e1",
+			expectedRes:            "v20200717-cf288082e1",
+			expectError:            false,
+			prefix:                 "gcr.io/k8s-prow/",
+		},
+		{
+			description:            "a valid response will return the correct tag with other prefixes in the same file",
+			upstreamURL:            "auto",
+			upstreamServerResponse: "other random garbage\n image: gcr.io/k8s-other/deck:v22222222-cf288082e1\n image: gcr.io/k8s-prow/deck:v20200717-cf288082e1\n     image: gcr.io/k8s-another/deck:v11111111-cf288082e1",
+			expectedRes:            "v20200717-cf288082e1",
+			expectError:            false,
+			prefix:                 "gcr.io/k8s-prow/",
+		},
+	}
+
+	for _, tc := range cases {
+		t.Run(tc.description, func(t *testing.T) {
+			if tc.upstreamURL == "auto" {
+				// generate a test server so we can capture and inspect the request
+				testServer := httptest.NewServer(http.HandlerFunc(func(res http.ResponseWriter, req *http.Request) {
+					res.Write([]byte(tc.upstreamServerResponse))
+				}))
+				defer func() { testServer.Close() }()
+				tc.upstreamURL = testServer.URL
+			}
+
+			res, err := parseUpstreamImageVersion(tc.upstreamURL, tc.prefix)
+			if res != tc.expectedRes {
+				t.Errorf("The expected result %q != the actual result %q", tc.expectedRes, res)
+			}
+			if tc.expectError && err == nil {
+				t.Errorf("Expected to get an error but the result is nil")
+			}
+			if !tc.expectError && err != nil {
+				t.Errorf("Expected to not get an error but got one: %v", err)
+			}
+		})
+	}
+}
+
+func TestUpstreamImageVersionResolver(t *testing.T) {
+	prowProdFakeVersion := "v-prow-prod-version"
```

Comment:
a) these can all be consts, right?

```golang
const (
  prowProdFakeVersion = "v-prow-prod-version"
  prowStagingFakeVersion = "v-prow-staging-version"
  ...
)
```

b) While it can be more verbose, when it comes times to read and maintain unit tests IMO it tends to be easier if you put as much info entirely inside the table as as little as possible outside it. Although this can be an art. 

AKA it is much easie to read `expectedTargetTag: "v-prow-prod-version"` than it is to hunt down what `prowProdFakeVersion` is set to.

Reply:
done